src/publish.ts

```diff
@@ -0,0 +1,323 @@
+import path from 'path'
+import os from 'os'
+
+class Publish {
+  private token: string
+  private charmhubToken: string
+  private octokit
+  private workflowFile: string
+  private artifact
+  private workingDir: string
+  private resourceMapping: { [key: string]: string }
+  private workflowRunId: string
+  private identifier: string
+
+  constructor() {
+    this.token = core.getInput('github-token')
+    this.charmhubToken = core.getInput('charmhub-token')
+    this.octokit = github.getOctokit(this.token)
+    this.workflowFile = core.getInput('workflow-file')
+    this.workingDir = core.getInput('working-directory')
+    this.resourceMapping = JSON.parse(core.getInput('resource-mapping'))
+    this.workflowRunId = core.getInput('workflow-run-id')
+    this.identifier = core.getInput('identifier')
+    this.artifact = new DefaultArtifactClient()
+  }
+
+  async findWorkflowRunId(): Promise<number> {
+    if (this.workflowRunId !== '') {
+      return parseInt(this.workflowRunId)
+    }
+    const owner = github.context.repo.owner
+    const repo = github.context.repo.repo
+    const commit = await this.octokit.rest.git.getCommit({
+      owner,
+      repo,
+      commit_sha: github.context.sha
+    })
+    const tree = commit.data.tree.sha
+    const workflowResp = await this.octokit.rest.actions.getWorkflow({
+      owner,
+      repo,
+      workflow_id: this.workflowFile
+    })
+    if (workflowResp.status !== 200) {
+      throw new Error(
+        `failed to find integration workflow (${this.workflowFile}): status ${workflowResp.status}`
+      )
+    }
+    const runIter = this.octokit.paginate.iterator(
+      this.octokit.rest.actions.listWorkflowRuns,
+      {
+        owner,
+        repo,
+        workflow_id: workflowResp.data.id,
+        status: 'success',
+        created: `>${new Date(Date.now() - 7 * 24 * 60 * 60 * 1000).toISOString().slice(0, 10)}`
+      }
+    )
+    for await (const resp of runIter) {
+      if (resp.status !== 200) {
+        throw new Error(
+          `failed to find integration workflow run: status ${resp.status}`
+        )
+      }
+      for (const run of resp.data) {
+        if (run?.head_commit?.tree_id === tree) {
+          return run.id
+        }
+      }
+    }
+    throw new Error(
+      `Failed to find integration test workflow run on tree id (${tree}).` +
+        "Consider enabling the 'Require branches to be up to date before merging' setting to ensure that the integration tests are executed on the merged commit"
+    )
+  }
+
+  async run() {
+    try {
+      core.startGroup('retrieve image info')
+      const images = await this.getImages()
+      core.endGroup()
+      core.startGroup('retrieve charm info')
+      const {
+        name: charmName,
+        dir: charmDir,
+        files: charms
+      } = await this.getCharms()
+      core.endGroup()
+      core.info(
+        `start uploading image resources: ${JSON.stringify(Object.fromEntries([...images]))}`
```

Comment:
Out of curiosity, is there a reason why you're spreading the images?

Reply:
I am not very sure what spreading the images means here.

Reply:
Oh sorry i should have clarified, i was referring to the use of ... spread operator inside a list [...images]. Would it be redundant here?

Reply:
Yes, I think `...images` doesn't create an iterator like Python so I think the `[]` is necessary here.

Reply:
From what I understand, obect.fromEntries creates an object from list of key-value pair (list), would images already be in such format?